### PR TITLE
Event sourced [NOT FINISHED]

### PR DESCRIPTION
**Pull request by hallgren**
I started this branch to extract the event sourcing part from AggregateRoot. I think that the concept Aggregate Root from DDD has nothing to do with event sourcing and that the event sourcing part should stand on its own legs.

I have started to implement EventSourced that is a copy of AggregateRoot, and I will try to extract only the stuff that is needed in the new Module.

Happy for comments and code.

### CODE REVIEW

File: lib/sandthorn/event_sourced_base.rb [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+        end
+
+        self
+      end
+
+      alias :record_event :commit
+
+      module ClassMethods
+
+        @@event_sourced_attributes = []
```
Comment (jesjos) [inline lib/sandthorn/event_sourced_base.rb]: Is this use of class variables intended? Subclasses share the same class variables, so you can get very interesting bugs in this way. For example: if you create a subclass and then run `event_sourced_attributes` again, you will mutate what attributes the superclass will event source.

File: spec/event_sourced_spec.rb [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+          expect(subject).to match_array(ids)
+        end
+      end
+
+    end
+
+
+    describe "when making a change on a aggregate" do
+      let(:dirty_obejct) { 
+        o = DirtyClass.new
```
Comment (jesjos) [inline spec/event_sourced_spec.rb]: Superfluous assignment

File: spec/event_sourced_spec.rb [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+      end
+      it "sets the event store as a class level variable and returns it" do
+        klass.event_store(:other)
+        expect(klass.event_store).to eq(:other)
+      end
+    end
+
+    describe "when get all aggregates from DirtyClass" do
+      
+      before(:each) do
```
Comment (jesjos) [inline spec/event_sourced_spec.rb]: `:each` not needed

File: spec/event_sourced_spec.rb [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+        end
+      end
+
+
+    end
+
+    describe "event data" do
+
+      let(:dirty_obejct) { 
+        o = DirtyClass.new :name => "old_value", :sex => "hen"
```
Comment (jesjos) [inline spec/event_sourced_spec.rb]: Superfluous assignment

File: spec/event_sourced_spec.rb [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+        it "should include correct aggregates" do
+          expect(subject).to match_array(ids)
+        end
+      end
+
+    end
+
+
+    describe "when making a change on a aggregate" do
+      let(:dirty_obejct) {
```
Comment (jesjos) [inline spec/event_sourced_spec.rb]: Typo

File: spec/event_sourced_spec.rb [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+          dirty_obejct.change_sex "hen"
+          dirty_obejct.events.each do |event|
+            event[:event_args][:attribute_deltas].each do |attribute_delta|
+              expect(attribute_delta[:attribute_name]).not_to eql "sex"
+            end
+          end
+        end
+
+        it "should set old_value and new_value on sex change" do
+          dirty_obejct.change_sex "shemale"
```
Comment (jesjos) [inline spec/event_sourced_spec.rb]: Can we not use this term?

File: lib/sandthorn/event_sourced_base.rb [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+          )
+          @events = []
+          @originating_version = @current_event_version
+        end
+
+        self
+      end
+
+      def ==(other)
+        other.respond_to?(:id) && id == other.id
```
Comment (jesjos) [inline lib/sandthorn/event_sourced_base.rb]: I guess this is copied code, but I think we should change that the other 'thing' has the same `version` as well.

File: lib/sandthorn/event_sourced_base.rb [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+            current_aggregate_version = aggregate.originating_version
+            events.shift
+          else
+            aggregate = create_new_empty_aggregate
+          end
+
+          attributes = build_instance_vars_from_events events
+          current_aggregate_version = events.last[:aggregate_version] unless events.empty?
+          aggregate.send :clearevents_
+          aggregate.send :set_orginating_aggregate_version!, current_aggregate_version
```
Comment (jesjos) [inline lib/sandthorn/event_sourced_base.rb]: Typo `orginating` => `originating`

Comment (jesjos) [inline lib/sandthorn/event_sourced_base.rb]: If we stop talking about aggregates, this name should change.

File: lib/sandthorn/event_sourced_base.rb (hunk 1 above)
Comment (hallgren) [inline lib/sandthorn/event_sourced_base.rb]: Check commit f40b2b8

File: lib/sandthorn/event_sourced_base.rb (hunk 7 above)
Comment (hallgren) [inline lib/sandthorn/event_sourced_base.rb]: Check commit 6807491

**Pull request closed without merging**